### PR TITLE
[Community Tutorial] Detecting changes in Sentinel-1 imagery, part 2

### PR DESCRIPTION
**Pull request by mortcanty**
Preview link to the colab notebook

https://colab.research.google.com/drive/1-uPBiXykgFQC5xpqO19LxIcqAKcW6zm9?usp=sharing

### CODE REVIEW

**Comment by jdbcode** [comment]
Thanks for the PR @mortcanty! @gino-m and I will be working on the review - expect feedback within a week.

- [ ] Subject and editorial review - @jdbcode
- [ ] Production review for publishing - @gino-m

File: tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb
Comment (jdbcode) [inline tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb:1]: <p>This gives me a little anxiety. There is an assumption that there are two images. Part of me would like you to just define the two images explicitly, on the other hand I do like that you are showing people how to filter the collection. I think it would make me feel better if you extended the date range to all of August to better ensure there are at least two dates, then in the next section print the dates and say "let's select the first two images". The above text cell will need to be edited to reflect code changes.</p><p> </p><pre class="ql-syntax" spellcheck="false">im_coll = ee.ImageCollection('COPERNICUS/S1_GRD_FLOAT') \
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; .filterBounds(aoi) \
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; .filterDate(ee.Date('2020-08-01'),ee.Date('2020-09-01')) \
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; .filter(ee.Filter.eq('orbitProperties_pass', 'ASCENDING')) \
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; .filter(ee.Filter.eq('relativeOrbitNumber_start', 15)) \
&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; .sort('system:time_start')
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/222/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (jdbcode) [inline tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb:1]: <p>The 100 argument might be confusing. How about:</p><p> </p><pre class="ql-syntax" spellcheck="false">im_list = im_coll.toList(im_coll.size())
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/222/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (jdbcode) [inline tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb:1]: <p><em>matplotlib</em> : &gt; <em>matplotlib</em>:</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/222/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='22'/>

Comment (jdbcode) [inline tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb:1]: <p>In last paragraph: occured &gt; occurred</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/222/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='28'/>

Comment (jdbcode) [inline tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb:1]: <p>Suggestions to improve visual interpretation:</p><p> </p><ul><li>Use "Stamin Toner" as the default basemap and include OpenStreetMap as an option</li><li>Include the ratio image as a layer for added context</li><li>Change green to blue for those with green-red color blindness</li></ul><p> </p><pre class="ql-syntax" spellcheck="false">#&nbsp;Decision&nbsp;threshold&nbsp;alpha/2:
dt&nbsp;=&nbsp;f.ppf(0.0005,&nbsp;2*m,&nbsp;2*m)

#&nbsp;LRT&nbsp;statistics.
q1&nbsp;=&nbsp;im1.divide(im2)
q2&nbsp;=&nbsp;im2.divide(im1)

#&nbsp;Change&nbsp;map&nbsp;with&nbsp;0&nbsp;=&nbsp;no&nbsp;change,&nbsp;1&nbsp;=&nbsp;decrease,&nbsp;2&nbsp;=&nbsp;increase&nbsp;in&nbsp;intensity.
c_map&nbsp;=&nbsp;im1.multiply(0).where(q2.lt(dt),&nbsp;1)
c_map&nbsp;=&nbsp;c_map.where(q1.lt(dt),&nbsp;2)

#&nbsp;Mask&nbsp;no-change&nbsp;pixels.
c_map&nbsp;=&nbsp;c_map.updateMask(c_map.gt(0))

#&nbsp;Display&nbsp;map&nbsp;with&nbsp;red&nbsp;for&nbsp;increase&nbsp;and&nbsp;blue&nbsp;for&nbsp;decrease&nbsp;in&nbsp;intensity.
location&nbsp;=&nbsp;aoi.centroid().coordinates().getInfo()[::-1]
mp&nbsp;=&nbsp;folium.Map(location=location,&nbsp;zoom_start=13,&nbsp;height=800,&nbsp;width=1000,&nbsp;tiles='Stamen&nbsp;Toner')
folium.TileLayer('OpenStreetMap').add_to(mp)
mp.add_ee_layer(ratio,&nbsp;{'min':&nbsp;0,'max':&nbsp;20,&nbsp;'palette':&nbsp;['black','white']},&nbsp;'Ratio')
mp.add_ee_layer(c_map,&nbsp;{'min':0,'max':2,'palette':&nbsp;['black','blue','red']},&nbsp;'Change&nbsp;Map')
mp.add_child(folium.LayerControl())

display(mp)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/222/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='36'/>

Comment (jdbcode) [inline tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb:1]: <p>restriced &gt; restricted</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/222/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='40'/>

Comment (jdbcode) [inline tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb:1]: <ul><li>Remove space at start of function def</li><li>Break up last line:</li></ul><pre class="ql-syntax" spellcheck="false">m2logQ&nbsp;=&nbsp;det(im1).log().add(det(im2).log()).subtract(
&nbsp;&nbsp;&nbsp;&nbsp;det(im1.add(im2)).log().multiply(2)).add(4*np.log(2)).multiply(-2*m) 
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/222/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='41'/>

Comment (jdbcode) [inline tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb:1]: <p>Similar to above comment on image selection, I would feel better if you extended the date filter to all of August and all of September - saying something like "we'll select the first image acquired in August and September"</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/222/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='47'/>

Comment (jdbcode) [inline tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb:1]: <p>Remove <code>import time</code>  since it was imported previously.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/222/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='49'/>

Comment (jdbcode) [inline tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb:1]: <p>Suggest to break up long line like this:</p><p> </p><pre class="ql-syntax" spellcheck="false">#&nbsp;The&nbsp;observed&nbsp;test&nbsp;statistic&nbsp;image&nbsp;-2logq.
m2logq&nbsp;=&nbsp;det(im1).log().add(det(im2).log()).subtract(
&nbsp;&nbsp;&nbsp;&nbsp;det(im1.add(im2)).log().multiply(2)).add(4*np.log(2)).multiply(-2*m)
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/222/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='51'/>

Comment (jdbcode) [inline tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb:1]: <p>agricutural &gt; agricultural</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/222/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='55'/>

Comment (jdbcode) [inline tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb:1]: <ul><li>As above, suggest to change green color to blue here and in previous text cell</li><li>Break up long line:</li></ul><pre class="ql-syntax" spellcheck="false">mp.add_ee_layer(c_map.updateMask(c_map.gt(0)),
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;{'min':&nbsp;0,'max':&nbsp;3,&nbsp;'palette':&nbsp;['black','yellow','blue','red']},&nbsp;'c_map')
</pre>
<br/>

 _Reply via <a href='https://app.reviewnb.com/google/earthengine-community/pull/222/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='61'/>

**Comment by jdbcode** [comment]
Hi @mortcanty - again, excellent tutorial - thanks! I've included some minor editorial notes and provided a recommendation to change how collections are filtered, and how the change layers are visualized in Folium, among other small nits. Please also check that there is a space between function arguments in all code cells e.g. `fun(arg1,arg2)` > `fun(arg1, arg2)` and note that we have a style preference for adding space between math operators e.g. `1+1` > `1 + 1`.

I did my review before your most recent commit ("minor changes for consistency with Part 3") and the changes are not obvious because track changes shows that the entire file is new. Could you please point out any major differences?

You'll find it easiest to see review comments in [ReviewNB](https://app.reviewnb.com/google/earthengine-community/pull/222/discussion/).

I'll look forward to seeing the changes.

**Comment by mortcanty** [comment]
@jdbcode Thanks for the careful reading and good suggestions. I've taken them all into account (I hope), also corrected some minor blips, e.g., $P$ -> _P_ in the text boxes, and added another reference for the Loewner order. The changes you missed previously were similar. 

BTW, my workflow is 
  -  edit CoLab notebook on Google Drive
  -  download to my local cloned repo, overwriting earlier version
  -  commit change and push to the forked repo
Is there a better way?

On to Part 3, which will take a bit longer :-)

File: tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb
Comment (jdbcode) [reply tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb]: <p>Done - verified.</p>

Comment (jdbcode) [reply tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb]: <p>Done - verified.</p>

Comment (jdbcode) [reply tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb]: <p>Done - verified.</p>

Comment (jdbcode) [reply tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb]: <p>Done - verified.</p>

Comment (jdbcode) [reply tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb]: <p>Done - verified.</p>

Comment (jdbcode) [reply tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb]: <p>Done - verified.</p>

Comment (jdbcode) [reply tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb]: <p>Done - verified.</p>

Comment (jdbcode) [reply tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb]: <p>Done - verified.</p>

Comment (jdbcode) [reply tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb]: <p>Done - verified.</p>

Comment (jdbcode) [reply tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb]: <p>Done - verified.</p>

Comment (jdbcode) [reply tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb]: <p>Done - verified.</p>

**Comment by jdbcode** [comment]
Thanks for the changes @mortcanty. I'll pass it over to @gino-m for a final look.

RE edit/commit process: your process is the same as mine. You can edit the .ipynb JSON directly, which provides precise change tracking, but IMO is too hard to do and is error prone.

I forgot to mention previously, my reviews have not included a thorough analysis of the equations and theory - I'm trusting your expertise 🤓

Also, I like the Judge Roy Bean analogy!

**Comment by mortcanty** [comment]
That's fine with me. The Roy Bean "test" really does drive it home. A statistics Prof passed it on to be years ago.

**Comment by jdbcode** [comment]
Hi @mortcanty I see that there is a conflict that cropped up after part 1 of the tutorial was submitted. There is some entanglement due to directly editing the master branch for tutorial part one, submitting a PR, and simultaneously creating a new branch from the edited master branch for the part 2 branch for this PR. Can you try to resolve it using the commands that GitHub suggests for resolution at the bottom of this interface. If you can't easily resolve it, maybe we close this PR and have you submit a fresh one after you have synced your master branch to the google/earthengine-community master.
@gino-m any thoughts on this?

**Comment by mortcanty** [comment]
@gino-m The Resolve Conflicts button says "conflicts too complex to resolve in the web editor". Which means they are too complex for @mortcanty too. Can I proceed as follows

- close this PR
- delete this forked repository completely, 
- fork the GEE community repo again
- checkout a new branch for Part 2
- submit a new PR
 or is there a better way?

**Comment by gino-m** [comment]
> @gino-m The Resolve Conflicts button says "conflicts too complex to resolve in the web editor". Which means they are too complex for @mortcanty too. Can I proceed as follows
> 
> * close this PR
> * delete this forked repository completely,
> * fork the GEE community repo again
> * checkout a new branch for Part 2
> * submit a new PR
>   or is there a better way?

Hi @mortcanty, I may be able to save you some time - I'll try to pull locally and push a merge myself. I don't quite understand why there are conflicts with `tutorials/detecting-changes-in-sentinel-1-imagery-pt-2/tutorial.ipynb
`; since it hasn't been merged into master yet, there should be no content to conflict with. I'll take a quick look.

In the meantime, could you kindly remove Part 1 from the PR?

**Comment by gino-m** [comment]
> In the meantime, could you kindly remove Part 1 from the PR?

Please disregard my previous comment about Part 1. It seems I can't push to your branch, so I'll send you a PR that you can approve and merge to resolve the conflicts now.

**Comment by mortcanty** [comment]
Sorry, too fast, I just removed Part 1. Hope I haven't made things worse

**Comment by gino-m** [comment]
> Sorry, too fast, I just removed Part 1. Hope I haven't made things worse

That was quick! 🚀  No problem at all, I created https://github.com/mortcanty/earthengine-community/pull/3 to fix the two remaining issues. Once those are in, I'll try to import the tutorial to prepare it for publication. Thanks for your patience!

**Review by gino-m — APPROVED** [review]
LGTM! Approving so we can start moving through the internal publication process. Stay tuned!

**Comment by gino-m** [comment]
Hi @mortcanty, my sincere apologies for the delay on getting this published. We realized the tutorial would be more complete if we could include the contents on the output cells in the published version, but our existing Earth Engine docs do not yet support that out of the box. I've been researching several options for doing this, but a working solution is a few weeks out. I'm currently pushing to get this tutorial published without output for now, and to then republish it once we've added support for the various types of output cells used in the notebook. Would this be acceptable for you?

**Comment by mortcanty** [comment]
That's fine with me, and I'd appreciate having it out there.

**Comment by gino-m** [comment]
@mortcanty Sounds good. We're aiming to publish it today. I'll post an update here once it's live!

**Comment by jdbcode** [comment]
@mortcanty The second part of your tutorial has been published. You can find it here: https://developers.google.com/earth-engine/tutorials/community/detecting-changes-in-sentinel-1-imagery-pt-2

**Comment by mortcanty** [comment]
@gino-m @jdbcode  Many thanks. I'll announce it on the developers newsgroup.

**Comment by jdbcode** [comment]
Thanks for announcing on the developer forum, and thanks for the valuable contribution to folks knowledge of SAR and Earth Engine!

FYI: I posted an announcement to Twitter: https://twitter.com/jstnbraaten/status/1321585672063381504

**Comment by mortcanty** [comment]
@jdbcode Thanks for the Tweet. Maybe a silly question, but are these community tutorials maintainable? That is, are updates in light of new developments/improvements possible or are they frozen once published (like a textbook).

**Comment by gino-m** [comment]
Hi @mortcanty. Good question.. if you have additions or edits you'd like to make, you can submit a new Pull Request with the modified file and once approved,we'll republish it. Hope that helps!

**Pull request merged**
